### PR TITLE
feat: guide the user when no api found

### PR DESCRIPTION
**Pull request by olizilla**
Inform the user when we can't connect to the API and give them the option to set a custom api address.

screenshot 2018-07-06 17 25 51

Early WIP on #711 

License: MIT
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

**Comment by hacdias** [comment]
Still missing:

- [ ] Redirect
- [ ] Reload config

**Comment by hacdias** [comment]
Requires:

- [ ] https://github.com/ipfs-shipyard/ipfs-redux-bundle/pull/1

**Comment by hacdias** [comment]
@olizilla I've made some changes. Please take a look to [ipfs-redux-bundle](https://github.com/ipfs-shipyard/ipfs-redux-bundle/pull/1).

- If we can't connect to an instance, it redirects to `/welcome`.
- It won't let you open any other page if there isn't a valid IPFS instance.
	- This makes our navigation tests fail on CI because there isn't an IPFS instance running. Suggestions? We could start a js-ipfs instance so it connects and does everything else.
- Cache! So, you set an API address, reload the page, and it's still the same address.
- You can always go to `/welcome` and change the API url.

**Comment by hacdias** [comment]
@olizilla 

> Redirecting the url when ipfsReady is false feels like something that we should set up in redux.

It is on `redirects` bundle.

> We should catch the condition that the welcome page bundle has loaded, but IPFS isn't ready and there is no error. It can just show a loading spinner on the welcome in that case.

Done.

## When failed to connect.

image

## After being successfully connected

image

**Comment by olizilla** [comment]
@hacdias To complete this challenge please appease the CI gods. The tests should pass and the test coverage should not go down. Of note the navigation test that is failing was really a place holder for more exciting e2e tests. Do you accept this mission?

**Comment by hacdias** [comment]
For this PR I can add tests with and without a node running. We can then think about more tests haha. Sounds good @olizilla?

**Comment by olizilla** [comment]
@hacdias 💯

File: test/e2e/navigation.test.js [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+  page.waitForFunction(`document.title === 'Welcome to IPFS'`, {timeout: 5000})
+}, ms.minutes(1))
+
+it('Navigation test: node running', async () => {
   // Save 3s per run by using the initial page rather than creating a new one!
   // const page = await browser.newPage()
   const page = (await browser.pages())[0]
 
+  await page.evaluate(() => {
+    window.ipfs = node
```
Comment (olizilla) [inline test/e2e/navigation.test.js]: @hacdias ok so, this is a rad idea, but I don't think it'll work. With the e2e tests things get complicated, as the test is running in jest in node, and firing up a browser instance with it's own seperate runtime. I don't think we'll ever be able to create an ipfs instance in a node runtime and smoosh it into a browser without very weird things happening.

Much more minor, the api for `page.evaluate` requires you to pass any args you want to hand to the page, and they get magically transferred as args to the callback function when it is executed in the context of the page instance. Even so, like going through a black hole, an ipfs instance won't survive the trip.

One solution is to create a fake window.ipfs object with an id function that resolves with a test identity info, and then check for that. We don't need to test that js-ipfs works in the webui tests, so it's ok to stub it out.

**Comment by hacdias** [comment]
@olizilla made some updates and mocking but it seems to fail here: https://github.com/ipfs-shipyard/ipfs-webui/pull/715/commits/6a0f41dd6ff1caadef44c3e9a9c3910988415239#diff-0238dfb49a06ba796757bff9fcff0faaR41

There might be something on the files page that blocks the rendering? I can't see what it might be though. I tried adding empty functions for files.ls and files.stat but it didn't work.

**Comment by olizilla** [comment]
@hacdias i'll take a look at this now

**Comment by hacdias** [comment]
Thank you @olizilla

**Pull request merged**